src/value.rs

```diff
@@ -507,103 +507,114 @@ impl<'a, A: bigarray::BigarrayElt> OCaml<'a, bigarray::Array1<A>> {
+                ) -> Result<OCaml<'c, RetT>, OCamlException>
+                where
+                    $($argname: OCamlParam<'a, 'b, $rt, $ot>),+
+                {
+                    $(let $argname = $argname.to_rooted(cr);)*
+
+                    let result = unsafe { $method(self.get_raw(), $($argname.get_raw()),+) };
+                    if is_exception_result(result) {
+                        let ex = unsafe { OCamlException::of(extract_exception(result)) };
+                        Err(ex)
+                    } else {
+                        Ok(unsafe { OCaml::new(cr, result) })
+                    }
+                }
+            }
+        )*
+        $(
+            #[allow(clippy::too_many_arguments)]
+            #[allow(non_camel_case_types)]
+            impl<'c, $($ot2,)* RetT> BoxRoot<fn($($ot2,)*) -> RetT> {
+                /// Calls the OCaml closure, converting the arguments to OCaml if necessary
+                pub fn try_call<'a, 'b: 'a, $($argname2),* $(,$rt2)* >(
+                    &self,
+                    cr: &'c mut OCamlRuntime,
+                    $($argname2: $argname2),*
+                ) -> Result<OCaml<'c, RetT>, OCamlException>
+                where
+                    $($argname2: OCamlParam<'a, 'b, $rt2, $ot2>),*
+                {
+                    $(let $argname2 = $argname2.to_rooted(cr);)*
+
+                    #[allow(unused_unsafe)]
+                    let mut args = unsafe {
+                        [$($argname2.get_raw()),*]
+                    };
+
+                    let result = unsafe { caml_callbackN_exn(self.get_raw(), args.len(), args.as_mut_ptr()) };
+                    if is_exception_result(result) {
+                        let ex = unsafe { OCamlException::of(extract_exception(result)) };
+                        Err(ex)
+                    } else {
+                        Ok(unsafe { OCaml::new(cr, result) })
+                    }
+                }
+            }
+        )*
     }
 }
 
-impl_closure!(
-    caml_callback_exn,
-    (arg1: O1T->R1T) -> RetT);
-impl_closure!(
-    caml_callback2_exn,
-    (arg1: O1T->R1T, arg2: O2T->R2T) -> RetT);
-impl_closure!(
-    caml_callback3_exn,
-    (arg1: O1T->R1T, arg2: O2T->R2T, arg3: O3T->R3T) -> RetT);
-impl_closure!(
-    caml_callbackN_exn,
-    (arg1: O1T->R1T, arg2: O2T->R2T, arg3: O3T->R3T, arg4: O4T->R4T) -> RetT);
-impl_closure!(
-    caml_callbackN_exn,
-    (arg1: O1T->R1T, arg2: O2T->R2T, arg3: O3T->R3T, arg4: O4T->R4T, arg5: O5T->R5T) -> RetT);
-impl_closure!(
-    caml_callbackN_exn,
-    (arg1: O1T->R1T, arg2: O2T->R2T, arg3: O3T->R3T, arg4: O4T->R4T, arg5: O5T->R5T, arg6: O6T->R6T) -> RetT);
+try_call_impl! {
+    { caml_callback_exn, ((arg1: OCaml1 Rust1)) },
+    { caml_callback2_exn, ((arg1: OCaml1 Rust1), (arg2: OCaml2 Rust2)) },
+    { caml_callback3_exn, ((arg1: OCaml1 Rust1), (arg2: OCaml2 Rust2), (arg3: OCaml3 Rust3)) },
+    NPARAMS:
+    {}, // Call without arguments
```

Comment:
@sebastiencs I will review this later, but just one quick comment on this one. Functions in OCaml always take at least one argument (and in the case of functions that really don't use their input, that argument is just the unit value `()`).

Reply:
@tizoc  Ah ok, I've removed that one